libs/sdk-ui-dashboard/src/model/state/_infra/queryService.ts

```diff
@@ -0,0 +1,344 @@
+
+    /**
+     * Removes cache entry by cache key
+     */
+    remove: QueryCacheReducer<TQuery, TResult, string>;
+
+    /**
+     * Removes all entries from cache.
+     */
+    removeAll: QueryCacheReducer<TQuery, TResult, void>;
+};
+
+/**
+ * This is a specialization of toolkit's CaseReducerActions to correctly type the action creators.
+ */
+type QueryActions<TQuery extends IDashboardQuery, TResult> = CaseReducerActions<
+    AllQueryCacheReducers<TQuery, TResult>
+>;
+
+/**
+ * Contains all essentials of query cache that can be integrated into redux store and redux sagas.
+ */
+type QueryCache<TQuery extends IDashboardQuery<TResult>, TResult> = {
+    /**
+     * A name to use as key in _queryCache part of the redux state.
+     */
+    cacheName: string;
+
+    /**
+     * Cache's slice reducer. This needs to be integrated into the dashboard store.
+     */
+    reducer: Reducer<EntityState<QueryCacheEntry<TQuery, TResult>>>;
+
+    /**
+     * Cache's action creators.
+     */
+    actions: QueryActions<TQuery, TResult>;
+
+    /**
+     * Factory for selectors to get cache entry by its identifier.
+     *
+     * Important: the use-case for this selector is purely internal. It is intended to be used by the internal
+     * sagas that add caching layer on top of the actual query processing saga.
+     *
+     * @param id - cache key
+     */
+    selectById: (
+        id: EntityId,
+    ) => OutputSelector<
+        DashboardState,
+        QueryCacheEntryResult<TResult> | undefined,
+        (res: DashboardState) => QueryCacheEntryResult<TResult> | undefined
+    >;
+
+    /**
+     * Factory for selectors that obtain query result stored in state. This is intended for consumption by presentational components.
+     *
+     * The selected result may be undefined - which means that query to obtain the result has not (yet) started:
+     * it has never even been dispatched or it is on its way to be processed. Once the query is running the status in the
+     * result will change to `loading` and then eventually to either `success` or `error`.
+     *
+     * @param query - query to get result of
+     * @internal
+     */
+    selectQueryResult: (
+        query: TQuery,
+    ) => OutputSelector<
+        DashboardState,
+        QueryCacheEntryResult<TResult> | undefined,
+        (res: DashboardState) => QueryCacheEntryResult<TResult> | undefined
+    >;
+};
+
+/**
+ * Given query name, this function will create name to use for the slice that will hold the cache with query results.
+ *
+ * For slice name, the function drops the "GDC.DASH/QUERY." prefix and then the rest is camel-cased with the word 'Cache' appended at the end.
+ *
+ * @param queryName - query name in format GDC.DASH/QUERY.SOME.THING
+ */
+export function createSliceNameForQueryCache(queryName: string): string {
+    const withoutPrefix = queryName.split("/")[1];
+    invariant(
+        withoutPrefix,
```

Comment:
The validation here is looser than what the invariant says. Is that intentional?